### DISCUSSION TITLE
Cascading soft-delete: best practice?

### DISCUSSION THREAD

**Discussion by teuneboon** [sqlalchemy/sqlalchemy#7846] [Usage Questions] [answered]
For an application I'm developing one of the requirements was having soft-deletes. I've done something similar before, and that worked fine, but it used lazy loading of relationships for the cascade part(which was not the best idea for performance anyway), and right now I switched to async SQLAlchemy where I couldn't find a way to do lazy loading anymore. The current implementation looks something like this(simplified):

```python
class Deletable:
    deleted = Column(TINYINT(1), default=0, nullable=False)
    
    def delete(self, cascade=False):
        if cascade and self.deleted != 1:
            self.deleted = 2
        elif self.deleted != 2:
            self.deleted = 1
        
        self.delete_children()
            
    def delete_children(self):
        pass


class Car(Deletable):
    doors = relationship(Door)  # (this is a list)
    wheels = relationship(Wheel)  # (this is a list)
    engine = relationship(Engine)
    
    def delete_children(self):
        for door in self.doors:
            door.delete(True)
        
        for wheel in self.wheels:
            wheel.delete(True)
            
        self.engine.delete(True)
        
        
class Door(Deletable):
    handle = relationship(Handle)
    
    def delete_children(self):
        self.handle.delete(True)
        

class Handle(Deletable):
    pass
```

When all relationships were lazy I would just call `car.delete()` and all doors/wheels and doorhandles would be loaded in separate queries and also get their "deleted" column set to 2(to indicate that it's been deleted through a cascade instead of directly, to make undeleting possible). Inefficient but it works.

Now, with SQLAlchemy 2 syntax because I went async, when I want to delete a Car I first have to query the Car like this:
```python
car = await session.get(Car, car_id, options=[joinedload(Car.doors).joinedload(Door.handle), joinedload(Car.wheels), joinedload(Car.engine)])
car.delete()
```
for it to work. So the code that calls the delete needs to know about all relationships of `Car`. I tried to fix this issue by adding the following classmethod to Car:
```python
@classmethod
def options_for_delete():
    return [joinedload(Car.doors).joinedload(Door.handle), joinedload(Car.wheels), joinedload(Car.engine)]
```

This way I can simply call
```python
car = await session.get(Car, car_id, options=Car.options_for_delete())
```
which is much more acceptable. The problem with THIS setup is that now, Car needs to know that Doors have a handle. And when looking into how to solve that issue I kind of came to the conclusion: I can't be the first to run into this issue and there has to be some simpler way to solve this. So my question is: what's best practice here? I'm at a fairly early stage of this project, so a big refactor is much easier now then it is later, and I'd like to do it right.

Some other stuff I considered:
- add lazy='joined' to the relationships of Car, not really an option since I don't want to load all relationships all the time
- add a options_for_delete() to Door as well, which I'd then combine somehow in the options_for_delete() of Car. But I couldn't quite figure out how to do this.

**Comment by zzzeek** [comment, marked as answer]
I would set up those deletes inside of before_flush().  do all the loading there and you can use the sync APIs even with the async connection.   Take a look at https://docs.sqlalchemy.org/en/14/orm/examples.html#module-examples.versioned_history for how before_flush can be set up with objects that follow particular interfaces as you are doing.

**Reply by teuneboon** [reply]
Thanks, that looks like what I'll need.

**Comment by mmzeynalli** [comment]
@teuneboon were you able to solve this efficiently?

**Reply by valq7711** [reply]
This problem can be solved efficiently by treating the car as an abstract assembly consisting of other sub-assemblies (e.g. doors) and parts (e.g. wheels), i.e. there should be just one table to store parent-child relationships, in which case all children can be soft-deleted using `WITH RECURSIVE`-query with one shot without loading children at all
